src/support/generation.c

```diff
@@ -109,7 +106,15 @@ void
 			/* Ensure we only read the value once. */
 			WT_ORDERED_READ(v, s->generations[which]);
 
-			if (v == 0 || generation <= v)
+			/*
+			 * The generation argument is newer than the limit. Wait
+			 * for threads in generations older than the argument
+			 * generation, threads in argument generations are OK.
+			 *
+			 * The thread's generation may be 0 (that is, not set),
+			 * and that's always less than the argument generation.
+			 */
+			if (v < generation)
 				break;
```

Comment:
Yeah, you're right, it's a little disconcerting that testing didn't fail.

I think the correct test is `v == 0 || v >= generation`, since generation is the newest generation, we're waiting for older generations to drain.

Whether we're testing for `N` or `N+1` might be a bit subtle for an API when (if?) other functions call `__wt_gen_drain()`. I can't off-hand think of anything better, let me know if you do.